File: src/CapySR.GameServer/Handlers/LineupHandlers.cs
using CapySR.GameServer.Game;
using CapySR.GameServer.Net;
using CapySR.GameServer.Scene;
using CapySR.Protocol;
using Microsoft.Extensions.Logging;

namespace CapySR.GameServer.Handlers;

// The client redraws the party bar from SyncLineupNotify and the overworld models from the
// scene, and it takes both as read only until the request it sent is answered. So every edit
// pushes the scene and the lineup first and closes with the response, or the new team does
// not show until something else nudges the UI.
[Handlers]
public static class LineupHandlers
{
    public static Task OnGetAllLineupData(PlayerSession session, GetAllLineupDataCsReq request)
    {
        var book = session.Player.Lineups;
        var response = new GetAllLineupDataScRsp { Retcode = 0, CurIndex = book.CurrentIndex };

        for (var i = 0; i < book.Squads.Count; i++)
        {
            response.LineupList.Add(session.BuildLineupInfo(book.Squads[i], ExtraLineupType.LineupNone, (uint)i));
        }

        return session.SendAsync(response);
    }

    public static Task OnGetCurLineupData(PlayerSession session, GetCurLineupDataCsReq request) =>
        session.SendAsync(new GetCurLineupDataScRsp { Retcode = 0, Lineup = session.BuildWorldLineup() });

    public static async Task OnReplaceLineup(PlayerSession session, ReplaceLineupCsReq request)
    {
        var book = session.Player.Lineups;
        var entries = request.LineupSlotList
            .Where(s => session.OwnsBase(s.Id))
            .Select(s => (s.Slot, s.Id))
            .ToList();

        Lineup target;
        bool isCurrent;

        if (request.ExtraLineupType != ExtraLineupType.LineupNone)
        {
            target = book.SetExtra(request.ExtraLineupType, entries.Select(e => e.Id));
            isCurrent = true;
        }
        else
        {
            target = book.Squad(request.Index);
            target.Replace(entries, request.LeaderSlot);
            isCurrent = !book.InExtraLineup && request.Index == book.CurrentIndex;
        }

        // never leave the walking team empty; put the roster's first back
        if (target.IsEmpty && ReferenceEquals(target, book.Current))
        {
            target.Fill(session.DefaultLineup());
        }

        session.Player.Touch();
        session.LogTeam($"replace #{request.Index}", target);

        await session.PushTeamAsync(isCurrent);
        await session.SendAsync(new ReplaceLineupScRsp { Retcode = 0 });
    }

    public static async Task OnJoinLineup(PlayerSession session, JoinLineupCsReq request)
    {
        var target = session.Target(request.ExtraLineupType, request.Index);
        var avatarId = request.OPCJCDKMJEI;

        var retcode = !session.OwnsBase(avatarId)
            ? Retcode.RetLineupAvatarNotExist
            : target.Join(avatarId, request.Slot);

        session.LogTeam($"join {avatarId} -> slot {request.Slot} ({retcode})", target);

        if (retcode == Retcode.RetSucc)
        {
            session.Player.Touch();
            await session.PushTeamAsync(session.IsCurrent(request.ExtraLineupType, request.Index));
        }

        await session.SendAsync(new JoinLineupScRsp { Retcode = (uint)retcode });
    }

    public static async Task OnQuitLineup(PlayerSession session, QuitLineupCsReq request)
    {
        var target = session.Target(request.ExtraLineupType, request.Index);
        var retcode = target.Quit(request.BaseAvatarId);

        session.LogTeam($"quit {request.BaseAvatarId} ({retcode})", target);

        if (retcode == Retcode.RetSucc)
        {
            session.Player.Touch();
            await session.PushTeamAsync(session.IsCurrent(request.ExtraLineupType, request.Index));
        }

        await session.SendAsync(new QuitLineupScRsp { Retcode = (uint)retcode });
    }

    public static async Task OnSwapLineup(PlayerSession session, SwapLineupCsReq request)
    {
        var target = session.Target(request.ExtraLineupType, request.Index);
        var retcode = target.Swap(request.MPIEGFCNBEN, request.ENBIBCPPLDC);

        session.LogTeam($"swap {request.MPIEGFCNBEN} <-> {request.ENBIBCPPLDC} ({retcode})", target);

        if (retcode == Retcode.RetSucc)
        {
            session.Player.Touch();
            await session.PushTeamAsync(session.IsCurrent(request.ExtraLineupType, request.Index));
        }

        await session.SendAsync(new SwapLineupScRsp { Retcode = (uint)retcode });
    }

    public static async Task OnChangeLineupLeader(PlayerSession session, ChangeLineupLeaderCsReq request)
    {
        var lineup = session.Player.Lineups.Current;
        var retcode = lineup.SetLeader(request.Slot);

        session.LogTeam($"leader -> slot {request.Slot} ({retcode})", lineup);

        if (retcode == Retcode.RetSucc)
        {
            session.Player.Touch();

            if (session.Scene.Info is { } scene)
            {
                scene.LeaderEntityId = SceneBuilder.ActorEntityId(session.World.Data.BaseAvatarId(lineup.LeaderAvatarId));
            }

            await session.SyncLineupAsync();
        }

        await session.SendAsync(new ChangeLineupLeaderScRsp
        {
            Retcode = (uint)retcode,
            Slot = request.Slot,
        });
    }

    public static async Task OnSwitchLineupIndex(PlayerSession session, SwitchLineupIndexCsReq request)
    {
        var book = session.Player.Lineups;
        var retcode = book.InExtraLineup ? Retcode.RetLineupInvalidIndex : book.SetCurrent(request.Index);

        session.LogTeam($"switch -> #{request.Index} ({retcode})", book.Current);

        if (retcode == Retcode.RetSucc)
        {
            session.Player.Touch();
            await session.PushTeamAsync(isCurrent: true);
        }

        await session.SendAsync(new SwitchLineupIndexScRsp
        {
            Retcode = (uint)retcode,
            Index = request.Index,
        });
    }

    public static Task OnSetLineupName(PlayerSession session, SetLineupNameCsReq request)
    {
        var book = session.Player.Lineups;

        if (request.Index >= LineupBook.SquadCount)
        {
            return session.SendAsync(new SetLineupNameScRsp { Retcode = (uint)Retcode.RetLineupInvalidIndex });
        }

        book.Squad(request.Index).Name = request.Name;
        session.Player.Touch();

        return session.SendAsync(new SetLineupNameScRsp
        {
            Retcode = 0,
            Index = request.Index,
            Name = request.Name,
        });
    }

    // every owned character with its current health, for the team picker
    public static Task OnGetLineupAvatarData(PlayerSession session, GetLineupAvatarDataCsReq request)
    {
        var world = session.World;
        var response = new GetLineupAvatarDataScRsp { Retcode = 0 };

        foreach (var baseId in world.OwnedAvatarIds.Select(world.Data.BaseAvatarId).Distinct())
        {
            var path = Roster.ResolvePath(world, session.Player, baseId);

            response.AvatarDataList.Add(new LineupAvatarData
            {
                Id = baseId,
                AvatarType = AvatarType.AvatarFormalType,
                Hp = session.Player.StateOf(path, world.Player).Hp,
            });
        }

        return session.SendAsync(response);
    }

    private static async Task PushTeamAsync(this PlayerSession session, bool isCurrent)
    {
        if (isCurrent)
        {
            await session.RefreshSceneActorsAsync();
        }

        await session.SyncLineupAsync();
    }

    private static Lineup Target(this PlayerSession session, ExtraLineupType type, uint index)
    {
        var book = session.Player.Lineups;

        return type != ExtraLineupType.LineupNone
            ? book.Extra(type) ?? book.SetExtra(type, [], activate: false)
            : book.Squad(index);
    }

    private static bool IsCurrent(this PlayerSession session, ExtraLineupType type, uint index)
    {
        var book = session.Player.Lineups;

        return type != ExtraLineupType.LineupNone
            ? book.ExtraType == type
            : !book.InExtraLineup && index == book.CurrentIndex;
    }

    // a multipath character is addressed by its base id; owning any path counts
    public static bool OwnsBase(this PlayerSession session, uint avatarId)
    {
        if (avatarId == 0)
        {
            return false;
        }

        var world = session.World;
        return world.OwnedAvatarIds.Any(id => id == avatarId || world.Data.BaseAvatarId(id) == avatarId);
    }

    private static void LogTeam(this PlayerSession session, string what, Lineup lineup)
    {
        session.Logger.LogInformation("lineup {What}: [{Team}] leader {Leader}",
            what, string.Join(", ", lineup.Slots.Select(id => id == 0 ? "-" : id.ToString())), lineup.LeaderAvatarId);
    }
}

File: src/game/lineup_handlers.cpp
#include <algorithm>

#include "core/logger.h"
#include "game/handlers.h"
#include "game/notify.h"
#include "game/player.h"
#include "game/scene.h"
#include "net/cmd_ids.h"
#include "net/handler.h"
#include "net/session.h"
#include "proto/gen/protos.h"

namespace game {
namespace {

Player* playerOf(net::Session& session, const char* what) {
    Player* player = session.player();
    if (player == nullptr) logging::warn("game", "{} before login", what);
    return player;
}

void onGetCurLineupData(net::Session& session, const proto::GetCurLineupDataCsReq&) {
    Player* player = playerOf(session, "GetCurLineupData");
    if (player == nullptr) return;

    proto::GetCurLineupDataScRsp rsp;
    rsp.retcode = 0;
    rsp.lineup = scene::lineupInfo(*player);
    session.send(cmd::GetCurLineupDataScRsp, rsp);
}

void onGetAllLineupData(net::Session& session, const proto::GetAllLineupDataCsReq&) {
    Player* player = playerOf(session, "GetAllLineupData");
    if (player == nullptr) return;
    LineupBook& book = player->lineups();

    proto::GetAllLineupDataScRsp rsp;
    rsp.retcode = 0;
    rsp.cur_index = book.curIndex();
    for (uint32_t index = 0; index < kSquadCount; ++index) {
        rsp.lineup_list.push_back(scene::lineupInfo(*player, index));
    }
    session.send(cmd::GetAllLineupDataScRsp, rsp);
}

void onJoinLineup(net::Session& session, const proto::JoinLineupCsReq& req) {
    Player* player = playerOf(session, "JoinLineup");
    if (player == nullptr) return;

    // Field 1 is the base avatar id; the dump never got a readable name for it.
    uint32_t avatarId = req.OPCJCDKMJEI;
    if (!player->lineups().join(req.index, req.slot, avatarId)) {
        logging::debug("game", "could not put {} in squad {} slot {}", avatarId, req.index,
                       req.slot);
    }
    player->save();
    if (req.index == player->lineups().curIndex()) notify::lineupChanged(session, *player);
    session.sendEmpty(cmd::JoinLineupScRsp);
}

void onQuitLineup(net::Session& session, const proto::QuitLineupCsReq& req) {
    Player* player = playerOf(session, "QuitLineup");
    if (player == nullptr) return;

    player->lineups().quit(req.index, req.base_avatar_id);
    player->save();
    if (req.index == player->lineups().curIndex()) notify::lineupChanged(session, *player);
    session.sendEmpty(cmd::QuitLineupScRsp);
}

void onSwapLineup(net::Session& session, const proto::SwapLineupCsReq& req) {
    Player* player = playerOf(session, "SwapLineup");
    if (player == nullptr) return;

    // The two slot fields were never named in the dump; swapping is symmetric, so
    // which one is the source does not matter.
    player->lineups().swap(req.index, req.MPIEGFCNBEN, req.ENBIBCPPLDC);
    player->save();
    if (req.index == player->lineups().curIndex()) notify::lineupChanged(session, *player);
    session.sendEmpty(cmd::SwapLineupScRsp);
}

void onReplaceLineup(net::Session& session, const proto::ReplaceLineupCsReq& req) {
    Player* player = playerOf(session, "ReplaceLineup");
    if (player == nullptr) return;

    // The client does not always send the slots in order, and a zero id is a hole.
    std::vector<proto::LineupSlotData> slots = req.lineup_slot_list;
    std::sort(slots.begin(), slots.end(),
              [](const proto::LineupSlotData& a, const proto::LineupSlotData& b) {
                  return a.slot < b.slot;
              });
    std::vector<uint32_t> avatarIds;
    for (const proto::LineupSlotData& slot : slots) {
        if (slot.id != 0) avatarIds.push_back(slot.id);
    }
    player->lineups().replace(req.index, avatarIds, req.leader_slot);
    player->saveNow();
    if (req.index == player->lineups().curIndex()) notify::lineupChanged(session, *player);
    session.sendEmpty(cmd::ReplaceLineupScRsp);
}

void onChangeLineupLeader(net::Session& session, const proto::ChangeLineupLeaderCsReq& req) {
    Player* player = playerOf(session, "ChangeLineupLeader");
    if (player == nullptr) return;

    LineupBook& book = player->lineups();
    book.setLeaderSlot(book.curIndex(), req.slot);
    player->save();

    proto::ChangeLineupLeaderScRsp rsp;
    rsp.retcode = 0;
    rsp.slot = req.slot;
    session.send(cmd::ChangeLineupLeaderScRsp, rsp);
}

void onSetLineupName(net::Session& session, const proto::SetLineupNameCsReq& req) {
    Player* player = playerOf(session, "SetLineupName");
    if (player == nullptr) return;

    if (req.index < kSquadCount) player->lineups().squad(req.index).name = req.name;
    player->saveNow();

    proto::SetLineupNameScRsp rsp;
    rsp.retcode = 0;
    rsp.index = req.index;
    rsp.name = req.name;
    session.send(cmd::SetLineupNameScRsp, rsp);
}

void onSwitchLineupIndex(net::Session& session, const proto::SwitchLineupIndexCsReq& req) {
    Player* player = playerOf(session, "SwitchLineupIndex");
    if (player == nullptr) return;
    LineupBook& book = player->lineups();

    proto::SwitchLineupIndexScRsp rsp;
    rsp.index = req.index;
    if (req.index >= kSquadCount) {
        rsp.retcode = fail(proto::Retcode::RET_LINEUP_INVALID_INDEX);
    } else if (book.members(req.index).empty()) {
        // Answering nothing here would leave the client waiting on a request it
        // never sees completed.
        rsp.retcode = fail(proto::Retcode::RET_LINEUP_IS_EMPTY);
    } else {
        book.setCurIndex(req.index);
        player->saveNow();
        notify::lineupChanged(session, *player);
        rsp.retcode = 0;
    }
    session.send(cmd::SwitchLineupIndexScRsp, rsp);
}

void onGetLineupAvatarData(net::Session& session, const proto::GetLineupAvatarDataCsReq&) {
    Player* player = playerOf(session, "GetLineupAvatarData");
    if (player == nullptr) return;
    Roster roster = player->roster();

    // The team screen reads hp from here rather than from the lineup, so every owned
    // avatar has to be listed -- including the ones sitting on the bench.
    proto::GetLineupAvatarDataScRsp rsp;
    rsp.retcode = 0;
    for (uint32_t baseId : roster.baseAvatarIds()) {
        proto::LineupAvatarData data;
        data.id = baseId;
        data.hp = kFullHp;
        data.avatar_type = proto::AvatarType::AvatarType_AvatarFormalType;
        rsp.avatar_data_list.push_back(data);
    }
    session.send(cmd::GetLineupAvatarDataScRsp, rsp);
}

void onRecoverAllLineup(net::Session& session, const proto::RecoverAllLineupCsReq&) {
    Player* player = playerOf(session, "RecoverAllLineup");
    if (player == nullptr) return;

    // Nobody ever loses hp here, so this only has to push the party back so the client
    // redraws the bars it just emptied.
    proto::SyncLineupNotify sync;
    sync.lineup = scene::lineupInfo(*player);
    sync.reason_list.push_back(proto::SyncLineupReason::SyncLineupReason_SyncReasonMpAdd);
    session.send(cmd::SyncLineupNotify, sync);

    proto::RecoverAllLineupScRsp rsp;
    rsp.retcode = 0;
    session.send(cmd::RecoverAllLineupScRsp, rsp);
}

void onMarkPresetLineup(net::Session& session, const proto::MarkPresetLineupCsReq& req) {
    Player* player = playerOf(session, "MarkPresetLineup");
    if (player == nullptr) return;

    proto::SetTeamFavourite rsp;
    rsp.index = req.index;
    rsp._is_favourite = req._is_favourite;
    if (req.index >= kSquadCount) {
        rsp.retcode = fail(proto::Retcode::RET_LINEUP_INVALID_INDEX);
    } else {
        player->lineups().squad(req.index).favourite = req._is_favourite;
        player->saveNow();
        rsp.retcode = 0;
    }
    // The dump never names a MarkPresetLineupScRsp; 773 is it, under the name the
    // client's own symbols use.
    session.send(cmd::SetTeamFavourite, rsp);
}

}  // namespace

void registerLineupHandlers() {
    net::on<proto::GetCurLineupDataCsReq>(cmd::GetCurLineupDataCsReq, onGetCurLineupData);
    net::on<proto::GetAllLineupDataCsReq>(cmd::GetAllLineupDataCsReq, onGetAllLineupData);
    net::on<proto::JoinLineupCsReq>(cmd::JoinLineupCsReq, onJoinLineup);
    net::on<proto::QuitLineupCsReq>(cmd::QuitLineupCsReq, onQuitLineup);
    net::on<proto::SwapLineupCsReq>(cmd::SwapLineupCsReq, onSwapLineup);
    net::on<proto::ReplaceLineupCsReq>(cmd::ReplaceLineupCsReq, onReplaceLineup);
    net::on<proto::ChangeLineupLeaderCsReq>(cmd::ChangeLineupLeaderCsReq, onChangeLineupLeader);
    net::on<proto::SetLineupNameCsReq>(cmd::SetLineupNameCsReq, onSetLineupName);
    net::on<proto::SwitchLineupIndexCsReq>(cmd::SwitchLineupIndexCsReq, onSwitchLineupIndex);
    net::on<proto::GetLineupAvatarDataCsReq>(cmd::GetLineupAvatarDataCsReq, onGetLineupAvatarData);
    net::on<proto::RecoverAllLineupCsReq>(cmd::RecoverAllLineupCsReq, onRecoverAllLineup);
    net::on<proto::MarkPresetLineupCsReq>(cmd::MarkPresetLineupCsReq, onMarkPresetLineup);
}

}  // namespace game
